cdap-master/src/main/java/co/cask/cdap/data/tools/ReplicationStatus.java

```diff
@@ -0,0 +1,290 @@
+
+        if (nextPart.matches(Constants.ReplicationStatusTool.SHUTDOWN_TIME_ROW)) {
+          // get shutdown time from master cluster
+          if (clusterID.matches(masterCluster)) {
+            shutdownTime = getShutdownTime(result);
+          }
+          continue;
+        }
+
+        if (clusterID.matches(masterCluster)) {
+          long writeTime = getTimeFromResult(result, Constants.ReplicationStatusTool.WRITE_TIME);
+          updateWriteTime(regionMap, nextPart, writeTime);
+        } else {
+          long replicateTime = getTimeFromResult(result, Constants.ReplicationStatusTool.REPLICATE_TIME);
+          updateReplicateTime(regionMap, nextPart, replicateTime);
+        }
+      }
+    } catch (Exception e) {
+      LOG.error("Error while reading table:" + e);
+      throw Throwables.propagate(e);
+    } finally {
+      hTable.close();
+    }
+
+    // Make sure master shutdown time is present
+    if (shutdownTime == 0L) {
+      System.out.println("CDAP Shutdown Time not available."
+                           + " Please make sure CDAP has been shutdown on the Master Cluster.");
+      return;
+    }
+
+    //Check the map for all regions
+    for (Map.Entry<String, ArrayList<Long>> tupleEntry: regionMap.entrySet()) {
+      if (!isReplicationComplete(shutdownTime, tupleEntry.getValue().get(writeTime),
+                                 tupleEntry.getValue().get(replicateTime))) {
+        System.out.println("HBase Replication is NOT complete. Region:" + tupleEntry.getKey() + " behind by " +
+                             (tupleEntry.getValue().get(writeTime) - tupleEntry.getValue().get(replicateTime)) + "ms.");
+        return;
+      }
+    }
+
+    // If replication is complete for all regions.
+    System.out.println("HBase Replication is complete.");
+  }
+
+  private static byte[] getNextPart (ByteBuffer byteBuffer) {
+    int len = byteBuffer.getInt();
+    if (byteBuffer.remaining() < len) {
+      throw new BufferUnderflowException();
+    }
+    byte[] bytes = new byte[len];
+    byteBuffer.get(bytes, 0, len);
+    return bytes;
+  }
+
+  private static TableId getReplicationStateTableId(HBaseTableUtil tableUtil) throws IOException {
+    String tableName = hConf.get(Constants.ReplicationStatusTool.REPLICATION_STATE_TABLE_NAME);
+    String ns = hConf.get(Constants.ReplicationStatusTool.REPLICATION_STATE_TABLE_NAMESPACE);
+
+    TableId tableId = tableUtil.createHTableId(
+      (ns != null)
+        ? new NamespaceId(ns)
+        : new NamespaceId(Constants.ReplicationStatusTool.REPLICATION_STATE_TABLE_DEFAULT_NAMESPACE),
+      (tableName != null) ? tableName : Constants.ReplicationStatusTool.REPLICATION_STATE_TABLE_DEFAULT_NAME);
+
+    return tableId;
+  }
+
+  private static void updateWriteTime(HashMap<String, ArrayList<Long>> map, String region, long timestamp) {
+    ArrayList tuple = map.get(region);
+    if (tuple == null) {
+      tuple = new ArrayList();
+      tuple.add(writeTime, timestamp);
```

Comment:
ah I see. For something like this, it's better to just make an internal class that has 2 fields rather than use a List.